src/coreclr/tools/Common/TypeSystem/Common/AsyncMethodVariant.cs

```diff
@@ -0,0 +1,138 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Internal.JitInterface;
+
+
+namespace Internal.TypeSystem
+{
+    /// <summary>
+    /// Either the AsyncMethodImplVariant or AsyncMethodThunkVariant of a method marked .IsAsync.
+    /// </summary>
+    public partial class AsyncMethodVariant : MethodDelegator, IJitHashableOnly
+    {
+        private readonly AsyncMethodVariantFactory _factory;
+        private readonly AsyncMethodKind _asyncMethodKind;
+        private readonly int _jitVisibleHashCode;
+        private MethodSignature _asyncSignature;
+
+        public AsyncMethodVariant(MethodDesc wrappedMethod, AsyncMethodVariantFactory factory, AsyncMethodKind kind)
+            : base(wrappedMethod)
+        {
+            Debug.Assert(wrappedMethod.IsTaskReturning);
+            Debug.Assert(kind switch
+            {
+                AsyncMethodKind.AsyncVariantThunk => !wrappedMethod.IsAsync,
+                AsyncMethodKind.AsyncVariantImpl => wrappedMethod.IsAsync,
+                _ => false,
+            });
+            _factory = factory;
+            _asyncMethodKind = kind;
+            _jitVisibleHashCode = HashCode.Combine(wrappedMethod.GetHashCode(), 0x310bb74f);
+        }
+
+        public MethodDesc Target => _wrappedMethod;
+
+        public override AsyncMethodKind AsyncMethodKind => _asyncMethodKind;
+
+        public override MethodSignature Signature
+        {
+            get
+            {
+                return _asyncSignature ??= _wrappedMethod.Signature.CreateAsyncSignature();
```

Comment:
Inline `CreateAsyncSignature` to here? It doesn't feel like a `MethodSignature` concern. Do we need to do different things based on thunk kind (drop `Task` vs add `Task`).